Dump Several Map-Related NARCs

ravepossum opened the pull request:
First of all - my apologies for the size of this PR; it got way out of hand. I will summarize the relevant information and parts that most need review as best I can.

# This PR:
- Dumps the following NARCs:
  - The map prop model NARC (`build_model.narc`)
  - The map prop texture set NARC (`areabm_texset.narc`)
  - The map prop model set NARC (`area_build.narc`)
  - The map prop animation NARC (`bm_anime.narc`)
  - The lighting information NARC (`arealight.narc`)
  - The area data NARC (`area_data.narc`)
- Renames the `json2bin` tools folder to `jsoncnv` since it's now also used to create text files
- Changes the map prop model constants to be generated with `metang`
- Moves the map matrix folder from `res/field/maps/` to `res/field/`
- Changes the hardcoded constants for area data ID and matrix ID in `include/data/map_headers.h` to instead use naix constants

# Primary areas of review:
- The file tree structure of the dumped NARCs (and the field/map data in general):
```
res/
----field/
--------area_data/
--------lighting/
--------matrices/
--------maps/
------------data/ (binary info, this will be broken up further later for the future map editor)
------------texture_sets/
--------props/
------------animations/
------------model_sets/
------------models/
------------texture_sets/
```
- The python conversion scripts for each NARC dumped to a json format:
  - `tools/jsoncnv/area_data.py`
  - `tools/jsoncnv/lighting.py`
  - `tools/jsoncnv/prop_model_set.py`

- The build scripts for each NARC dumped:
  - `res/field/area_data/meson.build`
  - `res/field/lighting/meson.build`
  - `res/field/props/model_sets/meson.build`
  - `res/field/props/models/meson.build`
  - `res/field/props/texture_sets/meson.build`
  - `res/field/props/animations/meson.build`
  
- Performance impact (build speeds from clean using WSL2):
  - Before PR:
  ```
  Executed in   47.38 secs    fish           external
   usr time  433.21 secs  345.00 micros  433.21 secs
   sys time   74.30 secs    0.00 micros   74.30 secs
   
  Executed in   47.12 secs    fish           external
   usr time  430.18 secs    0.00 micros  430.18 secs
   sys time   72.85 secs  315.00 micros   72.85 secs
   
  Executed in   47.74 secs    fish           external
   usr time  440.47 secs  152.00 micros  440.47 secs
   sys time   76.74 secs  188.00 micros   76.74 secs
  ```
  
  - After PR:
  ```
  Executed in   47.99 secs    fish           external
   usr time  454.66 secs  152.00 micros  454.66 secs
   sys time   74.51 secs  187.00 micros   74.51 secs

  Executed in   48.95 secs    fish           external
   usr time  460.44 secs  150.00 micros  460.44 secs
   sys time   75.93 secs  185.00 micros   75.93 secs

  Executed in   49.07 secs    fish           external
   usr time  465.39 secs    0.00 micros  465.39 secs
   sys time   77.14 secs  319.00 micros   77.14 secs
  ```

- The format of each NARC that was dumped to a json format:

Area Data:
```json
  {
    "mapPropSet": "prop_model_set_000",
    "mapTextureSet": "map_texture_set_000",
    "lightingSet": "lighting_set_000",
    "dummy": 0
  }
  ```

 Lighting:
  ```json
  [
    {
        "endTime": 0,
        "lights": [
            {
                "enabled": true,
                "color": {
                    "red": 11,
                    "green": 11,
                    "blue": 16
                },
                "direction": {
                    "x": 0.463134765625,
                    "y": -0.87890625,
                    "z": -0.11376953125
                }
            },
            {
                "enabled": false,
                "color": {
                    "red": 0,
                    "green": 0,
                    "blue": 0
                },
                "direction": {
                    "x": 0.0,
                    "y": 0.0,
                    "z": 0.0
                }
            },
            {
                "enabled": true,
                "color": {
                    "red": 22,
                    "green": 18,
                    "blue": 0
                },
                "direction": {
                    "x": 0.0,
                    "y": 0.0,
                    "z": 1.0
                }
            },
            {
                "enabled": true,
                "color": {
                    "red": 31,
                    "green": 31,
                    "blue": 31
                },
                "direction": {
                    "x": 0.0,
                    "y": 0.0,
                    "z": 1.0
                }
            }
        ],
        "diffuseColor": {
            "red": 14,
            "green": 14,
            "blue": 16
        },
        "ambientColor": {
            "red": 10,
            "green": 10,
            "blue": 10
        },
        "specularColor": {
            "red": 14,
            "green": 14,
            "blue": 16
        },
        "emissionColor": {
            "red": 8,
            "green": 8,
            "blue": 11
        }
    },
    
    ... (more lighting settings follow in a similar format)
  ```
 
 Prop model sets:
  ```json
  {
    "mapPropModels": [
        "MAP_PROP_MODEL_193",
        "MAP_PROP_MODEL_476",
        "MAP_PROP_MODEL_477",
        "MAP_PROP_MODEL_478",
        "MAP_PROP_MODEL_039",
        "MAP_PROP_MODEL_040",
        "MAP_PROP_MODEL_GYM_DOOR",
        "MAP_PROP_MODEL_539",
        "MAP_PROP_MODEL_540",
        "MAP_PROP_MODEL_541",
        "MAP_PROP_MODEL_542",
        "MAP_PROP_MODEL_543"
    ]
  }
  ```

lhearachel commented on res/field/maps/texture_sets/meson.build:
```diff
@@ -1,4 +1,9 @@
-map_texture_set_files = copy_gen.process(files(
+map_texture_set_gen = generator(find_program('cp'),
```
**suggestion**: We already have `copy_gen` for this in `res/meson.build`.

ravepossum replied:
The reason I didn't use `copy_gen` is because it uses `@PLAINNAME@` for the output, and I wanted to just use `@BASENAME@`.

lhearachel replied:
Out of curiosity: for what reason?

ravepossum replied:
To make the naix constants not include the file extension.

lhearachel commented on res/field/props/models/meson.build:
```diff
@@ -0,0 +1,618 @@
+    output: '@BASENAME@'
+)
+
+prop_model_files = files(
```
**suggestion**: I would rather we build this list dynamically from `generated/map_prop_models.txt`, like we do for `res/pokemon` and `res/trainer`. This also removes the need to maintain `map_prop_models.txt` and `prop_model_sets.order` separately.

lhearachel commented on tools/jsoncnv/area_data.py:
```diff
@@ -0,0 +1,41 @@
+
+from convert import parse_naix, u16
+
+def try_get_dict_value(naix_dict: dict, json_key: str, data: dict) -> int:
+    if data[json_key] in naix_dict:
+        return naix_dict[data[json_key]] 
+    else:
+        raise Exception(f'{json_key}: {data[json_key]} not found.')
```
**suggestion**: Leverage `KeyError` here to simplify the function:

```suggestion
def try_get_dict_value(naix_dict: dict, json_key: str, data: dict) -> int:
    return naix_dict[data[json_key]]
```

You could also emit a different error message by wrapping the `parse_area_data` call that ultimately surfaces this `KeyError` to the end-user in a `try-except` structure.

ravepossum replied:
I was thinking an error of the format:

`mapTextureSet: my_new_map not found.`

would be more helpful to a layperson end user than:

`KeyError: my_new_map`

But I don't feel that strongly about it. I could just go with the `try-except` approach you mentioned.

lhearachel replied:
The `KeyError` object itself contains this information already, is my thought, and we can have even greater control over the output message if we wrap the actual exception-point.

lhearachel commented on tools/jsoncnv/convert.py:
```diff
@@ -91,3 +96,9 @@ def derive_data_flags(party: list[dict]) -> TrainerDataFlags:
 
     return TrainerDataFlags(has_moves, has_items)
+
+def parse_naix(filepath: str) -> dict:
+    with open(filepath) as file:
+        matches = re.findall(r'#define (\S+) (\d+)', file.read())
+        naix_dict = dict((x, int(y)) for x, y in matches)
+        return naix_dict
```
**suggestion**: Add `import pathlib` to the top of the file, then:

```suggestion
def parse_naix(filepath: pathlib.Path) -> dict[str, int]:
    with open(filepath, "r") as file:
        return {
            match[1]: int(match[2])
            for match in re.finditer(r"#define (\w+) (\d+)", file.read())
        }
```

This:

1. Condenses the function into a comprehension (which later-versions often interpret into more efficient bytecode).
2. Uses `finditer` to eliminate some overhead to convert the matches into a `list` of string-tuples.

There is an option here, too, to use named-groups:

```py
r"#define (?P<ident>\w+) (?P<value>\d+)" # access with `match["ident"]` and `match["value"]`
```

But I will leave that to your discretion.

lhearachel reviewed (commented):
This mostly looks great. Some comments below refer to minor improvements.

ravepossum commented:
- Force push 1 addresses feedback related to `area_data.py`.
- Force push 2 makes the naming scheme for both documented and undocumented map prop models consistent so it's easier to use generated constants for the names.
- Force push 3 is a silly missed EOF newline.

That should be all the current feedback addressed.

ravepossum commented:
Knocking back to draft to reevaluate the build process for some of the NARCs.
